### PR TITLE
Add code unit prefix, less than, and general sorting

### LINKED ISSUE #55: Define string sorting by "code unit order"
Background: https://github.com/whatwg/url/pull/199

We want to update specs to be unambiguous about this, so I think infra is a good place to define it. It should include some examples (similar to the URLSearchParams WPTs).

### PR DESCRIPTION
**Pull request by domenic**
Closes #55.

Will be needed for Import Maps.

***
<a href="https://whatpr.org/infra/248.html" title="Last updated on May 9, 2019, 3:53 PM UTC (66540e1)">Preview</a> | <a href="https://whatpr.org/infra/248/8970e9f...66540e1.html" title="Last updated on May 9, 2019, 3:53 PM UTC (66540e1)">Diff</a>

### CODE REVIEW

File: infra.bs [hunk 1]
```diff
@@ -542,6 +543,74 @@ actually ends up representing <a lt="JavaScript string">JavaScript</a> and
...
+
+<ol>
+ <li><p>Let <var>i</var> be 0.
+
+ <li>
+  <p><a>While</a> true:
+
+  <ol>
+   <li><p>Let <var>aCodeUnit</var> be the <var>i</var>th <a>code unit</a> of <var>a</var>, or null
+   if <var>i</var> is greater than or equal to <var>a</var>'s <a for=string>length</a>.
```
Comment (annevk) [inline infra.bs]: How about applying https://github.com/whatwg/infra/issues/247#issuecomment-488227069 here and below as a start?

File: infra.bs [hunk 2]
```diff
@@ -542,6 +543,74 @@ actually ends up representing <a lt="JavaScript string">JavaScript</a> and
...
+
+   <li><p>If <var>aCodeUnit</var> is null and <var>bCodeUnit</var> is non-null, then return
+   true.
+
+   <li><p>If <var>aCodeUnit</var> is non-null and <var>bCodeUnit</var> is null, then return
+   false.
+
+   <li><p>Return false if <var>aCodeUnit</var> is different from <var>bCodeUnit</var>.
+
+   <li><p>Set <var>i</var> to <var>i</var> + 1.
```
Comment (annevk) [inline infra.bs]: I've used "Increment i by x" for this, but maybe this is clearer? Another thing to settle when we do numbers.

File: infra.bs [hunk 3]
```diff
@@ -542,6 +543,74 @@ actually ends up representing <a lt="JavaScript string">JavaScript</a> and
...
+<p>A <a>string</a> <var>a</var> is <dfn export>code unit less than</dfn> a <a>string</a>
+<var>b</var> if the following steps return true:
+
+<ol>
+ <li><p>If <var>b</var> is a <a>code unit prefix</a> of <var>a</var>, then return false.
+
+ <li><p>If <var>a</var> is a <a>code unit prefix</a> of <var>b</var>, then return true.
+
+ <li><p>Let <var>n</var> be the smallest index such that the <var>n</var>th <a>code unit</a> of
+ <var>a</var> is different from the <var>n</var>th code unit of <var>b</var>. (There must be such an
```
Comment (annevk) [inline infra.bs]: There is* (it's a statement of fact, not a requirement).

File: infra.bs [hunk 4]
```diff
@@ -542,6 +543,74 @@ actually ends up representing <a lt="JavaScript string">JavaScript</a> and
...
+ <li><p>If the <var>n</var>th code unit of <var>a</var> is less than the <var>n</var>th code unit of
+ <var>b</var>, then return true.
+
+ <li><p>Return false.
+</ol>
+
+<p class="note">This matches the ordering used by JavaScript's <code>&lt;</code> operator, and its
+{{Array/sort()}} method on an array of strings. This ordering compares the 16-bit code units in each
+string, producing a highly efficient, consistent, and deterministic sort order. The resulting
+ordering will not match any particular alphabet or lexoicographic order, particular for
```
Comment (annevk) [inline infra.bs]: lexicographic*

Comment (annevk) [reply infra.bs]: particularly* (or "in particular" I suppose)

File: infra.bs [hunk 5]
```diff
@@ -8,13 +8,14 @@ Translation: ja https://triple-underscore.github.io/infra-ja.html
 </pre>
 
 <pre class="anchors">
-urlPrefix: https://tc39.github.io/ecma262/; spec: ECMA-262;
+urlPrefix: https://tc39.github.io/ecma262/#; spec: ECMA-262;
```
Comment (annevk) [inline infra.bs]: Wow, this is bad.

File: infra.bs [hunk 6]
```diff
@@ -542,6 +543,74 @@ actually ends up representing <a lt="JavaScript string">JavaScript</a> and
...
+
+ <li><p>Return false.
+</ol>
+
+<p class="note">This matches the ordering used by JavaScript's <code>&lt;</code> operator, and its
+{{Array/sort()}} method on an array of strings. This ordering compares the 16-bit code units in each
+string, producing a highly efficient, consistent, and deterministic sort order. The resulting
+ordering will not match any particular alphabet or lexoicographic order, particular for
+<a>code points</a> represented by a surrogate pair. [[!ECMA-262]]
+
```
Comment (annevk) [inline infra.bs]: Only need a single newline here.

File: infra.bs (hunk 2 above)
Comment (domenic) [reply infra.bs]: "Increment" doesn't appear in Infra yet, so I'll leave it. But we do have an outstanding feature request, in #99.

File: infra.bs (hunk 3 above)
Comment (domenic) [reply infra.bs]: Good catch. I'll go with "There has to be".

File: infra.bs (hunk 5 above)
Comment (domenic) [reply infra.bs]: Bikeshed appears to do some sort of fixup; the links in the live document all work. I was just bothered by the inconsistency.

**Comment by domenic** [comment]
Nits fixed. My preferred course of action would be closing #55 and opening a new feature request to add a "byte less than" or "byte ASCII lowercase less than" operation for byte sequences. Or, perhaps now that this general framework is in place, those could just be defined in Fetch.

**Comment by annevk** [comment]
There was also code point order for CSSOM, but okay.

**Pull request merged**